gap/SemilinearMatrixGroups.gi

```diff
@@ -137,11 +137,86 @@ function(n, q, s)
     return MatrixGroupWithSize(F, Concatenation(AandB, [C, D]), size);
 end);
 
+# For a sesquilinear form beta on the vector space GF(q ^ s) ^ (d / s) given 
+# by the Gram matrix <form>, we can interpret any u, v in GF(q) ^ d as elements of 
+# GF(q ^ s) ^ (d / s) via the standard basis. Taking the trace of beta(u, v)
+# yields a sesquilinear form on the vector space GF(q) ^ d.
+#
+# Takes a unitary form and returns a unitary form if <type> = "U-U"; takes a
+# unitary form and returns a symplectic form if <type> = "U-S"; takes a
+# symplectic form and returns a symplectic form if <type> = "S-S". 
+BindGlobal("TakeTraceOfSesquilinearForm",
+function(form, q, s, type)
+    local F, d, newForm, B, rootOfq, i, j, eiOverGFqsEntry, eiOverGFqsIndex,
+    ejOverGFqsEntry, ejOverGFqsIndex;
+
+    if not type in ["S-S", "U-S", "U-U"] then
+        ErrorNoReturn("<type> must be one of 'S-S', 'U-S' or 'U-U'");
+    fi;
+    if type = "U-U" and not IsInt(Sqrt(q)) then
+        ErrorNoReturn("<q> must be a square if <type> = 'U-U'");
+    fi;
+    if type = "U-S" and not s = 2 then
+        ErrorNoReturn("<s> must be two if <type> = 'U-S'");
+    fi;
+
+    F := GF(q);
+    d := s * Size(form);
+    newForm := NullMat(d, d, F);
+    B := CanonicalBasis(AsField(F, GF(q ^ s)));
+
+    if type = "U-U" then
+        rootOfq := Sqrt(q);
+    fi;
+
+    for i in [1..d] do
+        # The basis vector ei of GF(q) ^ d corresponds to a vector in 
+        # GF(q ^ s) ^ (d / s) with the entry eiOverGFqsEntry in the 
+        # eiOverGFqsIndex-th component.
+        eiOverGFqsEntry := B[(i - 1) mod s + 1];
+        eiOverGFqsIndex := QuoInt(i - 1, s) + 1;
+
+        for j in [1..i] do
+            ejOverGFqsEntry := B[(j - 1) mod s + 1];
+            ejOverGFqsIndex := QuoInt(j - 1, s) + 1;
+
+            if type = "S-S" then
+                newForm[i, j] := TraceOverFiniteField(eiOverGFqsEntry
+                                                            * form[eiOverGFqsIndex, ejOverGFqsIndex] 
+                                                            * ejOverGFqsEntry, 
+                                                      q, s);
+            elif type = "U-S" then
+                newForm[i, j] := TraceOverFiniteField(eiOverGFqsEntry
+                                                            * form[eiOverGFqsIndex, ejOverGFqsIndex] 
+                                                            * ejOverGFqsEntry ^ q, 
+                                                      q, s);
+            else
+                newForm[i, j] := TraceOverFiniteField(eiOverGFqsEntry 
+                                                            * form[eiOverGFqsIndex, ejOverGFqsIndex]
+                                                            * ejOverGFqsEntry ^ (rootOfq ^ s), 
+                                                      q, s);
+            fi;
```

Comment:
How about adding a local variable `exp` which is set to `1`, `q` or `rootOfq ^ s` at the start, depending on `type`; and another local variable `tmp`; and then this code can be simplified to the following IMHO more readable variant
```suggestion
            tmp := eiOverGFqsEntry * form[eiOverGFqsIndex, ejOverGFqsIndex] * ejOverGFqsEntry ^ exp;
            newForm[i, j] := TraceOverFiniteField(tmp, q, s);
```